org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/ProjectsManager.java

```diff
@@ -417,4 +428,36 @@ public boolean setAutoBuilding(boolean enable) throws CoreException {
 		return changed;
 	}
 
+	public List<FileSystemWatcher> registerWatchers() {
+		logInfo(">> registerFeature 'workspace/didChangeWatchedFiles'");
+		if (preferenceManager.getClientPreferences().isWorkspaceChangeWatchedFilesDynamicRegistered()) {
+			Set<String> sources = new HashSet<>();
+			try {
+				IProject[] projects = ResourcesPlugin.getWorkspace().getRoot().getProjects();
+				for (IProject project : projects) {
+					IJavaProject javaProject = JavaCore.create(project);
+					if (javaProject != null && javaProject.exists()) {
+						IClasspathEntry[] classpath = javaProject.getRawClasspath();
+						for (IClasspathEntry entry : classpath) {
+							if (entry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
```

Comment:
should we skip watching derived sources?

Reply:
will this cause a conflict with clients already registering changes on `**/*.java`?

Reply:
It won't.